dashboard/src/app/actions/verification.ts

```diff
@@ -1,18 +1,64 @@
-"use server";
-
-import { withDashboardAuthContext } from "@/auth/auth-actions";
-import { AuthContext } from "@/entities/authContext";
-import { checkTrackingDataExists } from "@/services/verification";
-
-export const verifyTrackingInstallation = withDashboardAuthContext(
-  async (ctx: AuthContext): Promise<boolean> => {
-    const { siteId } = ctx;
-    
-    try {
-      return await checkTrackingDataExists(siteId);
-    } catch (error) {
-      console.error('Verification failed:', error);
-      return false;
+'use server';
+
+import { withDashboardAuthContext } from '@/auth/auth-actions';
+import { AuthContext } from '@/entities/authContext';
+import { checkTrackingDataExists } from '@/services/verification';
+import { checkRateLimit } from '@/services/verification.service';
+import {
+  SendVerificationEmailData,
+  SendVerificationEmailSchema,
+  VerifyEmailData,
+  VerifyEmailSchema,
+  VerificationResult,
+} from '@/entities/verification';
+import { sendVerificationEmail, verifyEmail } from '@/services/verification.service';
+
+export const verifyTrackingInstallation = withDashboardAuthContext(async (ctx: AuthContext): Promise<boolean> => {
+  const { siteId } = ctx;
+
+  try {
+    return await checkTrackingDataExists(siteId);
+  } catch (error) {
+    console.error('Verification failed:', error);
+    return false;
+  }
+});
+
+export async function verifyEmailAction(data: VerifyEmailData): Promise<VerificationResult> {
+  try {
+    const validatedData = VerifyEmailSchema.parse(data);
+    return await verifyEmail(validatedData);
+  } catch (error) {
+    console.error('Verify email action error:', error);
+    return {
+      success: false,
+      error: 'Failed to verify email, please request a new verification email and try again.',
+    };
+  }
+}
+
+export async function resendVerificationEmailAction(formData: SendVerificationEmailData) {
+  try {
+    const validatedData = SendVerificationEmailSchema.parse(formData);
+    const { email } = validatedData;
+
+    const rateLimitCheck = await checkRateLimit(email);
+
+    if (!rateLimitCheck.allowed && rateLimitCheck.nextAllowedAt) {
+      const waitTime = Math.ceil((rateLimitCheck.nextAllowedAt?.getTime() - Date.now()) / 60000);
```

Comment:
Quibble: The variable `rateLimitCheck.nextAllowedAt` is always defined here:
```suggestion
      const waitTime = Math.ceil((rateLimitCheck.nextAllowedAt.getTime() - Date.now()) / 60000);
```